src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+            var updatedFilters = GetUpdatedFilters(sortedList, filterResults, filters, filterText);
+
+            if (bestFilterResult != null)
+            {
+                // Only hard select this result if it's a prefix match
+                // We need to do this so that
+                // * deleting and retyping a dot in a member access does not change the
+                //   text that originally appeared before the dot
+                // * deleting through a word from the end keeps that word selected
+                // This also preserves the behavior the VB had through Dev12.
+                var hardSelect = bestFilterResult.Value.CompletionItem.FilterText.StartsWith(filterText, StringComparison.CurrentCultureIgnoreCase);
+
+                return new EditorCompletion.FilteredCompletionModel(highlightedList, filteredItems.IndexOf(bestFilterResult.Value.CompletionItem), updatedFilters, hardSelect ? EditorCompletion.UpdateSelectionHint.NoChange : EditorCompletion.UpdateSelectionHint.SoftSelected, centerSelection: true, uniqueItem: null);
+            }
+            else
+            {
+                return new EditorCompletion.FilteredCompletionModel(highlightedList, 0, updatedFilters, EditorCompletion.UpdateSelectionHint.SoftSelected, centerSelection: true, uniqueItem: null);
+            }
+        }
+
+        private bool IsBetterDeletionMatch(FilterResult result1, FilterResult result2)
+        {
+            var item1 = result1.CompletionItem;
+            var item2 = result2.CompletionItem;
+
+            var prefixLength1 = item1.FilterText.GetCaseInsensitivePrefixLength(result1.FilterText);
+            var prefixLength2 = item2.FilterText.GetCaseInsensitivePrefixLength(result2.FilterText);
+
+            // Prefer the item that matches a longer prefix of the filter text.
+            if (prefixLength1 > prefixLength2)
+            {
+                return true;
+            }
+
+            if (prefixLength1 == prefixLength2)
+            {
+                // If the lengths are the same, prefer the one with the higher match priority.
+                // But only if it's an item that would have been hard selected.  We don't want
+                // to aggressively select an item that was only going to be softly offered.
+
+                var item1Priority = CalculatePriority(item1);
+                var item2Priority = CalculatePriority(item2);
+
+                if (item1Priority > item2Priority)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private int CalculatePriority(EditorCompletion.CompletionItem item)
+            => GetOrCreateRoslynItem(item)?.Rules?.SelectionBehavior != CompletionItemSelectionBehavior.HardSelection
+                    ? MatchPriority.Default
+                    : GetMatchPriority(item);
+
+        private EditorCompletion.FilteredCompletionModel HandleNormalFiltering(
```

Comment:
functions were reordered from the original file.  why?

Reply:
restored. thank you!

---
In reply to: [222087228](https://github.com/dotnet/roslyn/pull/29016#discussion_r222087228) [](ancestors = 222087228)